### PR TITLE
Core & Multiple modules: activity controls

### PR DESCRIPTION
**Pull request by dgirardi**
## Type of change

- [x] Feature

## Description of change

This is an implementation of the [activity controls infrastructure](https://github.com/prebid/Prebid.js/issues/9546), and a refactoring of many of the current control mechanisms:

 - storage access control (`bidderSettings.*.storageAllowed`, `deviceAccess`) are now routed through the `accessDevice` activity
 - user sync controls (`filterSettings`) go through `syncUser`
 - all the checks done by the GDPR enforcement module now use activities (in additions to the above, they also affect `fetchBids` and `enrichUids`)
 - there are completely new checks for the `transmit*` and `enrich*` activites, for both bid adapters and RTD modules

## Other information

Closes https://github.com/prebid/Prebid.js/issues/9546
Documentations coming soon!

### CODE REVIEW

**Review by robertrmartinez — APPROVED** [review]
Just couple minor comments.

Beast of a PR but very very cool!

I tested it very thoroughly using mock CMP's, all the different Activities, FPD transmit and updates and it all seems to work as expected.

One minor comment which we spoke in slack about is that for storage for bidders the default is NOT ALLOW.

And there is no easy way with Activity Set Config to flip this

This does not work for example:

```
                allowActivities: {
                    accessDevice: {
                        default: true,
                    }
                }
```

Because there is special bidder handling, bidder settings still denies bidders from storage.

So a workaround is like so:

```
                    accessDevice: {
                        rules: [
                            {
                                condition: ({componentType, componentName}) => {
                                    return true
                                },
                                allow: true
                            }
                        ]
                    }
```
To allow all access to storage (or of course using standard bidderSettings.)

Which I think is fine.

Another thing we spoke about offline is moving the `cfg.js` logic (handling of activity setConfig stuff) to an optional module.

Lastly, When using Activity controls, many console logs come up and clutter the console. Especially when adapters use storage. We spoke a bit on some ways to suppress them but not sure there is a nice clean way besides a debounce which may not be worth doing. Users can just filter out `Activity` in their consoles if needed really. `-Activity` will do this in chrome.

Anyways, this is really really well done. Thanks @dgirardi . Great stuff

File: src/activities/rules.js [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    return registry[activity] = registry[activity] || [];
+  }
+
+  function runRule(activity, name, rule, params) {
+    let res;
+    try {
+      res = rule(params);
+    } catch (e) {
+      logger.logError(`Exception in rule ${name} for '${activity}'`, e);
+      res = {allow: false, reason: e};
```
[the remark this answers is not in the document]
Comment (dgirardi) [reply src/activities/rules.js:18]: correct, my reasoning is when it throws the pub clearly meant to do something else, so failing loudly seems the best strategy.

**Comment by dgirardi** [comment]
`condition` can be omitted if the rule is always active (referring to the workaround above), e.g.

```
setConfig({
   allowActivities: {
     accessDevice: {
        rules: [
            {
                allow: true
            }
        ]
      }
   }
})
```

I do plan to update this very soon with configuration logic extracted into a module and cleaner logging (yes it can be filtered out, but it'd bother me if I had to).

File: modules/userId/index.js [hunk 2]
```diff
@@ -940,24 +963,28 @@ function getValidSubmoduleConfigs(configRegistry, submoduleRegistry) {
...
-  if (coreStorage.cookiesAreEnabled()) {
-    storageTypes.push(COOKIE);
-    if (coreStorage.getCookie(PBJS_USER_ID_OPTOUT_NAME)) {
-      logInfo(`${MODULE_NAME} - opt-out cookie found, storage disabled`);
-      disabled = true;
-    }
+function canUseStorage(submodule) {
+  switch (submodule.config?.storage?.type) {
+    case LOCAL_STORAGE:
+      if (submodule.storageMgr.cookiesAreEnabled()) {
```
Comment (dgirardi) [inline modules/userId/index.js:969]: noting this for myself - this should be `localStorageIsEnabled`

**Review by pm-harshad-mane — APPROVED** [review]
Thanks @robertrmartinez for a very detailed review.
I did a high level review of this GRAND PR and it looks great to me 👍 
Thanks a lot @dgirardi for working this huge and highly important feature.

**Comment by pm-harshad-mane** [comment]
@dgirardi there seems an issue with E2E tests, can you please take a look?

**Pull request merged**